onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorAttention.cpp

```diff
@@ -0,0 +1,401 @@
+        std::array<int32_t, 3> strides {1, 1, 1};
+        DML_SLICE1_OPERATOR_DESC querySlicedOperatorDesc = {};
+        querySlicedOperatorDesc.InputTensor = &namedFirstGemmOutputTensorDesc;
+        querySlicedOperatorDesc.OutputTensor = &namedQuerySlicedInputTensorDesc;
+        querySlicedOperatorDesc.DimensionCount = gsl::narrow_cast<uint32_t>(querySlicedTensorShape.size());
+        querySlicedOperatorDesc.InputWindowOffsets = querySliceOffset.data();
+        querySlicedOperatorDesc.InputWindowSizes = sliceSize.data();
+        querySlicedOperatorDesc.InputWindowStrides = strides.data();
+        const DML_OPERATOR_DESC querySlicedDesc = { DML_OPERATOR_SLICE1, &querySlicedOperatorDesc };
+
+        DML_SLICE1_OPERATOR_DESC keySlicedOperatorDesc = {};
+        keySlicedOperatorDesc.InputTensor = &namedFirstGemmOutputTensorDesc;
+        keySlicedOperatorDesc.OutputTensor = &namedQuerySlicedInputTensorDesc;
+        keySlicedOperatorDesc.DimensionCount = gsl::narrow_cast<uint32_t>(querySlicedTensorShape.size());
+        keySlicedOperatorDesc.InputWindowOffsets = keySliceOffset.data();
+        keySlicedOperatorDesc.InputWindowSizes = sliceSize.data();
+        keySlicedOperatorDesc.InputWindowStrides = strides.data();
+        const DML_OPERATOR_DESC keySlicedDesc = { DML_OPERATOR_SLICE1, &keySlicedOperatorDesc };
+
+        DML_SLICE1_OPERATOR_DESC valueSlicedOperatorDesc = {};
+        valueSlicedOperatorDesc.InputTensor = &namedFirstGemmOutputTensorDesc;
+        valueSlicedOperatorDesc.OutputTensor = &namedQuerySlicedInputTensorDesc;
+        valueSlicedOperatorDesc.DimensionCount = gsl::narrow_cast<uint32_t>(querySlicedTensorShape.size());
+        valueSlicedOperatorDesc.InputWindowOffsets = valueSliceOffset.data();
+        valueSlicedOperatorDesc.InputWindowSizes = sliceSize.data();
+        valueSlicedOperatorDesc.InputWindowStrides = strides.data();
+        const DML_OPERATOR_DESC valueSlicedDesc = { DML_OPERATOR_SLICE1, &valueSlicedOperatorDesc};
+
+        TensorDesc castedMaskIndexTensorDesc = TensorDesc::ConstructDefaultTensorDesc(
+                                                                MLOperatorTensorDataType::Float, 
+                                                                desiredMaskIndexShape, 
+                                                                desiredMaskIndexShape
+                                                            );
+        DML_TENSOR_DESC namedCastedMaskIndexTensorDesc = castedMaskIndexTensorDesc.GetDmlDesc();
+
+        DML_CAST_OPERATOR_DESC castMaskIndexOperatorDesc = {};
+        castMaskIndexOperatorDesc.InputTensor = &inputDescs[3];
+        castMaskIndexOperatorDesc.OutputTensor = &namedCastedMaskIndexTensorDesc;
+        const DML_OPERATOR_DESC castMaskIndexDesc = {DML_OPERATOR_CAST, &castMaskIndexOperatorDesc};
+
+        // The attention fusion in ORT expects this to number to -10000. https://microsoft.visualstudio.com/OS/_workitems/edit/41864173
```

Comment:
>https://microsoft.visualstudio.com/OS/_workitems/edit/41864173 [](http://example.com/codeflow?start=73&length=62)

These links aren't accessible publicly. Also, it points to `Bring AP-TEST21 back online`.

Reply:
Thanks for pointing. Probably while pasting the link of github ORT repo, my ctrl+v might not copy the correct URL and I ended up pasting the AP-TEST21 bug URL. :joy: